hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/DualFileWriter.java

```diff
@@ -0,0 +1,253 @@
+    this.dualWriterEnabled = dualWriterEnabled;
+    this.newVersionBehavior = newVersionBehavior;
+    writers = new ArrayList<>(2);
+    initRowState();
+  }
+
+  private void initRowState() {
+    deleteFamily = null;
+    deleteFamilyVersionList.clear();
+    lastCell = null;
+  }
+
+  private void initColumnState() {
+    livePutCellCount = 0;
+    deleteColumn = null;
+    deleteColumnVersionList.clear();
+
+  }
+
+  private void addLiveVersion(Cell cell) throws IOException {
+    if (liveVersionWriter == null) {
+      liveVersionWriter = writerFactory.createWriter();
+      writers.add(liveVersionWriter);
+    }
+    liveVersionWriter.append(cell);
+  }
+
+  private void addHistoricalVersion(Cell cell) throws IOException {
+    if (historicalVersionWriter == null) {
+      historicalVersionWriter = writerFactory.createWriter();
+      writers.add(historicalVersionWriter);
+    }
+    historicalVersionWriter.append(cell);
+  }
+
+  private boolean isDeletedByDeleteFamily(Cell cell) {
+    return deleteFamily != null && (deleteFamily.getTimestamp() > cell.getTimestamp()
+      || (deleteFamily.getTimestamp() == cell.getTimestamp()
+        && (!newVersionBehavior || cell.getSequenceId() < deleteFamily.getSequenceId())));
+  }
+
+  private boolean isDeletedByDeleteFamilyVersion(Cell cell) {
+    for (Cell deleteFamilyVersion : deleteFamilyVersionList) {
+      if (
+        deleteFamilyVersion.getTimestamp() == cell.getTimestamp()
+          && (!newVersionBehavior || cell.getSequenceId() < deleteFamilyVersion.getSequenceId())
+      ) return true;
+    }
+    return false;
+  }
+
+  private boolean isDeletedByDeleteColumn(Cell cell) {
+    return deleteColumn != null && (deleteColumn.getTimestamp() > cell.getTimestamp()
+      || (deleteColumn.getTimestamp() == cell.getTimestamp()
+        && (!newVersionBehavior || cell.getSequenceId() < deleteColumn.getSequenceId())));
+  }
+
+  private boolean isDeletedByDeleteColumnVersion(Cell cell) {
+    for (Cell deleteColumnVersion : deleteColumnVersionList) {
+      if (
+        deleteColumnVersion.getTimestamp() == cell.getTimestamp()
+          && (!newVersionBehavior || cell.getSequenceId() < deleteColumnVersion.getSequenceId())
+      ) return true;
+    }
+    return false;
+  }
+
+  private boolean isDeleted(Cell cell) {
+    return isDeletedByDeleteFamily(cell) || isDeletedByDeleteColumn(cell)
+      || isDeletedByDeleteFamilyVersion(cell) || isDeletedByDeleteColumnVersion(cell);
+  }
+
+  private void appendCell(Cell cell) throws IOException {
+    if ((lastCell == null || !CellUtil.matchingColumn(lastCell, cell))) {
+      initColumnState();
+    }
+    if (cell.getType() == Cell.Type.DeleteFamily) {
+      if (deleteFamily == null) {
+        deleteFamily = cell;
+        addLiveVersion(cell);
+      } else {
+        addHistoricalVersion(cell);
+      }
+    } else if (cell.getType() == Cell.Type.DeleteFamilyVersion) {
+      if (!isDeletedByDeleteFamily(cell)) {
```

Comment:
Just in case, if DeleteFamily and DeleteFamilyVersion both have same timestamp (if at all possible), maybe this could be better handled with `if (!isDeletedByDeleteFamily(cell) && deleteFamily == null)`?

Although from data correctness, this is likely not an issue. It's just to protect DeleteFamilyVersion from going to live version file if DeleteFamily is already present in live version file with same timestamp. Maybe it's just a corner case.

Reply:
Dual file compaction is an optimization and does not guarantee that live version files will include only live cell versions. Since minor compaction does not see all the cell versions, dual file compaction may include cells that are not actually live since the decision is done based on a subset of files in the case of minor compaction.  I just wanted to make sure we have the same understanding on this.

If a DeleteFamily and DeleteFamilyVersion marker both have same timestamp and the DeleteFamilyVersion marker is not deleted/masked by the DeleteFamily marker then, the sequence id of the DeleteFamilyVersion marker must be higher than that of the DeleteFamily marker. In this case, we still need to track the DeleteFamilyVersion marker (i.e., need to add to the delete family version list). This tracking is necessary to make sure that the put cell with the same timestamp will be deleted by the DeleteFamilyVersion marker. And then we can append the DeleteFamilyVersion marker to the historical version file as you suggested. The same is also applicable to the Delete marker. I will make the changes for this in the next commit.

Reply:
Sounds good, that makes sense.